server/src/main/java/org/opensearch/transport/InboundPipeline.java

```diff
@@ -124,37 +128,42 @@ public void doHandleBytes(TcpChannel channel, ReleasableBytesReference reference
         statsTracker.markBytesRead(reference.length());
         pending.add(reference.retain());
 
-        final ArrayList<Object> fragments = fragmentList.get();
-        boolean continueHandling = true;
-
-        while (continueHandling && isClosed == false) {
-            boolean continueDecoding = true;
-            while (continueDecoding && pending.isEmpty() == false) {
-                try (ReleasableBytesReference toDecode = getPendingBytes()) {
-                    final int bytesDecoded = decoder.decode(toDecode, fragments::add);
-                    if (bytesDecoded != 0) {
-                        releasePendingBytes(bytesDecoded);
-                        if (fragments.isEmpty() == false && endOfMessage(fragments.get(fragments.size() - 1))) {
+        String incomingMessageProtocol = TcpTransport.determineTransportProtocol(reference);
```

Comment:
I would make a static map of protocol -> handler, and register supported protocols at startup (putting registering protobuf protocol behind a feature flag). 

This code then is `handler = map[incomingMessageProtocol]` that generically fails if no available protocol is defined.

Reply:
The transport protocol is determined though based on the prefix bytes in the header. The `InboundPipeline` here determines the protocol, deals with the bytes based on protocol and then ultimately calls the handler (native or protobuf).

Reply:
I am saying that `if (incomingMessageProtocol.equals(...)) { } else { }` should be replaced with:

```java
String incomingMessageProtocol = TcpTransport.determineTransportProtocol(reference);
InboundMessageProtocolHandler handler = this.transportProtocols.get(incomingMessageProtocol); // raises NoSuchProtocolException
```

Reply:
Makes sense! Done in the new commit.

Reply:
```
InboundMessageProtocolHandler handler = protocolBytesHandlers
    .stream()
    .filter(h -> h.canHandleBytes(reference))
    .findFirst()
    .orElseThrow(() ->)throw new IllegalStateException("No bytes handler found for protocol: " + incomingMessageProtocol));
```